### ISSUE TITLE
Option to resolve value of stage for serverless.ts

### ISSUE THREAD

**Issue by PatrykMilewski** [serverless/serverless#11939]
### Is there an existing issue for this?

- [X] I have searched existing issues, it hasn't been reported yet

### Use case description

When I'm using Serverless Framework using TypeScript for configuration, I cannot resolve stage name to the actual value

### Proposed solution (optional)

Provide helper function for resolving that:
```
import { getStage } from 'serverless/lib/utils';

const stage = getStage();
console.log('stage', stage);
// prints: stage: { "stageName" }
```

**Comment by medikoo** [comment]
Service stage, is a _service_ property, which can be accessed having `serverless` (framework service instance).

In `serverless.ts` you're simply in programmatic way construct a service configuration (there's no _service_ instance at this point), so technically you cannot access _resolved_ stage value there.

However you can refer to it via variable (`${sls:stage}`), which when setting up the service will be filled with resolved _stage_ value.

If you need to do something more sophisticated then you need to create a plugin, in which you can also define some custom variable for some more sophisticated configuration setup (custom variables have access to configuration property resolvers - so you can also access there resolved stage value - this logic is run at service setup stage)

**Comment by PatrykMilewski** [comment]
Yes, I understand all of those, however still I would like to be able to resolve the actual value of stage, so I can programatically decide on something, basing on the actual stage value. Using `${sls:stage}` won't help me much, because for example this function:
```
if ('${sls:stage}' !== 'dev') {
   return true;
}
```
will always return true

Comparing to CDK based frameworks/libs, doing something basing on the stage name, is quite common and really easy to achieve with a simple if statement. I believe it shouldn't be as hard as creating a new plugin for that

**Comment by medikoo** [comment]
@PatrykMilewski Plugins work in the context of the initialized service instance, and there's no service instance initialied when configuration is being defined.

We can agree there are the following processing phases (simplified):
1. Configuration definition - at this stage _stage_ cannot be resolved as we don't have a configuration fully defined
2. Configuration resolution (happens at service initialization) - at this stage _stage_ can be resolved conditionally (you can access it via e.g. custom variables resolver, assuming that properties that influence the _stage_ resolution are not hidden behind variables that cannot be resolved at this point)
3. Service execution - _Stage_ is fully resolved and accessible to loaded plugins

You're in context _1. Configuration definition_ (you're defining service configuration with TypeScript).

I see two ways:
1. Fully bulletproof - Properties of which value depend on resolved _stage_ need to be configured via variables, where in variables resolver, you access resolved stage (resolver functions will be run at Configuration resolution stage)
2. Assume that _stage_ will not be altered via configuration (no `provider.stage` property set), so assume that value of the stage will be either what's passed with  `--stage` CLI param (you need to resolve that param on your own) or `"dev"`

**Comment by colesiegel** [comment]
> * Fully bulletproof - Properties of which value depend on resolved _stage_ need to be configured via variables, where in variables resolver, you access resolved stage (resolver functions will be run at Configuration resolution stage)

Is there a typescript centric way to do this? [The docs](https://www.serverless.com/framework/docs/providers/aws/guide/variables#exporting-a-function) are all based on .js files, I'm having trouble understanding how I might use this in a strict typescript project with serverless.ts given the use of module.exports and lack of type support.

**Comment by medikoo** [comment]
> Is there a typescript centric way to do this? [The docs](https://www.serverless.com/framework/docs/providers/aws/guide/variables#exporting-a-function) are all based on .js files,

@colesiegel Serverless Framework since the beginning is (no compilation implied) a JavaScript project, and currently, there are no type bindings for any of its internal logic.

What indeed might be good to introduce is type bindings for the `Plugin` constructor, and all its properties as recognized by the Framework internals (and that'll include custom variable resolvers). That will definitely help TypeScript-heavy users that wish to construct plugins for the Framework.
If anyone is interested in providing this, we're definitely open for the PR.

Otherwise, you definitely can configure service plugins in TypeScript and in their context, fill the bindings for configured functions (as custom variables resolver).
